libraries/lib-audacity/LibAudacityApi.cpp

```diff
@@ -0,0 +1,219 @@
+#include "LibAudacityApi.h"
+#include "BasicUI.h"
+#include "CommandManager.h"
+#include "Effect.h"
+#include "EffectManager.h"
+#include "EffectPlugin.h"
+#include "PluginManager.h"
+#include "Project.h"
+#include "ProjectHistory.h"
+#include "ProjectRate.h"
+#include "TrackFocus.h"
+#include "ViewInfo.h"
+#include "Viewport.h"
+#include "WaveTrack.h"
+
+/// DoEffect() takes a PluginID and executes the associated effect.
+///
+/// At the moment flags are used only to indicate whether to prompt for
+//  parameters, whether to save the state to history and whether to allow
+/// 'Repeat Last Effect'.
+
+bool LibAudacity::DoEffect(
+   const PluginID& ID, AudacityProject& project, unsigned flags,
+   ShowEffectHostInterfaceCb showEffectHostInterfaceCb,
+   StopPlaybackCb stopPlaybackCb, SelectAllIfNoneCb selectAllIfNoneCb)
+{
+   auto& tracks = TrackList::Get(project);
+   auto& trackFactory = WaveTrackFactory::Get(project);
+   auto rate = ProjectRate::Get(project).GetRate();
+   auto& selectedRegion = ViewInfo::Get(project).selectedRegion;
+   auto& commandManager = CommandManager::Get(project);
+   auto& viewport = Viewport::Get(project);
+
+   const PluginDescriptor* plug = PluginManager::Get().GetPlugin(ID);
+
+   if (!plug || !PluginManager::IsPluginAvailable(*plug))
+   {
+      BasicUI::ShowMessageBox(
+         XO("This plugin could not be loaded.\nIt may have been deleted."),
+         BasicUI::MessageBoxOptions().Caption(XO("Plugin Error")));
+
+      return false;
+   }
+
+   EffectType type = plug->GetEffectType();
+
+   // Make sure there's no activity since the effect is about to be applied
+   // to the project's tracks.  Mainly for Apply during RTP, but also used
+   // for batch commands
+   if (flags & EffectManager::kConfigured)
+   {
+      stopPlaybackCb();
+      // Don't Select All if repeating Generator Effect
+      if (!(flags & EffectManager::kConfigured))
```

Comment:
I didn't pay attention. Was added in [this PR](https://github.com/audacity/audacity/pull/722). I can't see how this could be executed either, though. Remove ? Then we'd have one callback injection less.